src/anml/parameter/parameter.py

```diff
@@ -122,42 +122,157 @@ def design_mat(self, df: pd.DataFrame) -> np.ndarray:
         return np.ones((len(df), 1))
 
 
+@dataclass 
+class SplineLinearConstr:
+    """Constraints on spline derivatives. The general form is
+    lb <= Ax <= ub
+    where x is in some interval domain, and A can be 0th, 1st or 2nd order derivative matrix 
+    of the splines on evaluated at some discretization points in the domain.
+    A is not known at the initialization of this object, but will have dimension
+    `number of discretization points` by `number of spline basis`.
+    `lb` and `ub` are vectors of multiples of ones.
+
+    This type of constraints can be used to impose monotonicity and convexity constraints.
+    For instance, for splines defined on `[0, 5]`, one can specify monotonically decreasing on `[0,1]` with
+    `constr = SplineLinearConstr(x_domain=[0, 1], y_bounds=[-np.inf, 0.0], order=1)`, 
+    monotonically increasing on `[4, 5]` with 
+    `constr = SplineLinearConstr(x_domain=[4, 5], y_bounds=[0.0, np.inf], order=1)`,
+    and overall convexity with
+    `constr = SplineLinearConstr(x_domain=[0, 5], y_bounds=[0.0, np.inf], order=2)`.
+
+    Parameters
+    ----------
+    x_domain: List[float, float]
+        domain for x 
+    y_bounds: List[float, float]
+        bounds for y = Ax
+    order: int
+        order of the derivative
+
+    Raises
+    ------
+    ValueError
+        domain for x is not valid.
+    ValueError
+        bounds for y = Ax is not valid.
+    ValueError
+        invalid derivative order
+    """
+    x_domain: List[float, float]
+    y_bounds: List[float, float]
+    order: int
+
+    def __post_init__(self):
+        if self.x_domain[0] >= self.x_domain[1]:
+            raise ValueError('Domain must have positive length.')
+        if self.y_bounds[0] > self.y_bounds[1]:
+            raise ValueError('Lower bound cannot be greater than upper bound.')
+        if self.order < 0:
+            raise ValueError('Order of derivative must be nonnegative.')
+
+
 @dataclass
 class Spline(Variable):
-    """A spline variable.
+    """Spline variable.
+
+    Parameters
+    ----------
+    knots_type : str
+        type of knots. can only be 'frequency' or 'domain'
+    knots_num: int 
+        number of knots 
+    degree: int
+        degree of spines
+    l_linear: bool
+        whether left tail is linear
+    r_linear: bool
+        whether right tail is linear
+    include_intercept: bool
+        whether to include intercept in design matrix 
+    derivative_constr: List[`~anml.parameter.parameter.SplineLinearConstr`]
+        constraints on derivatives 
+    constr_grid_size: int
```

Comment:
feel like `constr_grid_size` could be attribute for `SplineLinearConstr` instead of `Spline` variable, so that different constraints could have different grid size.

Reply:
yes, right. in that case, do you mean that `grid_size` refers to the global domain or just domain of the constraint? e.g., for splines defined on `[0, 5]` and monotonicity on `[0,1]`, should we divide `[0, 5]` be divided into `constr_grid_size` or `[0, 1]` be divided that way?

Reply:
I think as long as each constraint has its own `grid_size`, no matter is the total number of the grid points or the grid density, it will be ok. Before my practice is to use total number of grid points, so that it is more explicit, but I can see the advantage of `grid_density` as well.